problem-solver/cxx/exampleModule/agents/PathFindingAgent.cpp

```diff
@@ -0,0 +1,161 @@
+  {
+    SC_AGENT_LOG_ERROR("Connector template argument is not found");
+    return action.FinishWithError();
+  }
+
+  if (!connectorWeightTemplateAddr.IsValid())
+  {
+    SC_AGENT_LOG_ERROR("Connector weight template argument is not found");
+    return action.FinishWithError();
+  }
+
+  if (startNode == endNode)
+  {
+    SC_AGENT_LOG_ERROR("Start and end nodes should not match");
+    return action.FinishWithError();
+  }
+
+  try
+  {
+    ConnectorTemplateInfo connectorTemplateInfo;
+    utils::TemplateUtils::getConnectorTemplateInfo(m_context, connectorTemplateAddr, connectorTemplateInfo);
+
+    WeightTemplateInfo weightTemplateInfo;
+    utils::TemplateUtils::getWeightTemplateInfo(m_context, connectorWeightTemplateAddr, weightTemplateInfo);
+
+    unsigned pathLength;
+    ScAddrVector path;
+
+    PathSearcher searcher(&m_context);
+    searcher.findPath(graph, startNode, endNode, connectorTemplateInfo, weightTemplateInfo, pathLength, path);
+
+    ScStructure const & result = formResult(path, pathLength, connectorTemplateInfo, weightTemplateInfo);
+    action.SetResult(result);
+  }
+  catch (ScException const & exception)
+  {
+    SC_AGENT_LOG_ERROR(exception.Message());
+    return action.FinishWithError();
+  }
+
+  return action.FinishSuccessfully();
+}
+
+ScStructure PathFindingAgent::formResult(
+    ScAddrVector const & path,
+    unsigned pathLength,
+    ConnectorTemplateInfo const & connectorTemplateInfo,
+    WeightTemplateInfo const & weightTemplateInfo) const
+{
+  ScStructure resultStructure = m_context.GenerateStructure();
+  ScStructure pathStructure = m_context.GenerateStructure();
+
+  for (auto it = path.begin() + 1; it != path.end(); ++it)
+  {
+    ScAddr const & first = *(it - 1);
+    ScAddr const & second = *it;
+
+    addConnectionIntoStructure(first, second, connectorTemplateInfo, pathStructure);
+  }
+
+  addPathWeightIntoStructure(pathStructure, weightTemplateInfo, pathLength, resultStructure);
+
+  return resultStructure;
+}
+
+void PathFindingAgent::addConnectionIntoStructure(
+    ScAddr const & first,
+    ScAddr const & second,
+    ConnectorTemplateInfo const & connectorTemplateInfo,
+    ScStructure & structure) const
+{
+  ScTemplateParams connectorTemplateParams;
+  connectorTemplateParams.Add(connectorTemplateInfo.connectorStartVariable, first);
+  connectorTemplateParams.Add(connectorTemplateInfo.connectorEndVariable, second);
+
+  ScTemplate connectorTemplate;
+  m_context.BuildTemplate(connectorTemplate, connectorTemplateInfo.templateAddr, connectorTemplateParams);
+
+  bool isFound = false;
+
+  m_context.SearchByTemplateInterruptibly(
+      connectorTemplate,
+      [&isFound, &structure](ScTemplateResultItem const & item) -> ScTemplateSearchRequest
+      {
+        isFound = true;
+
+        structure << item;
+
+        return ScTemplateSearchRequest::STOP;
+      });
+
+  if (!isFound)
+    SC_THROW_EXCEPTION(utils::ExceptionItemNotFound, "PathFindingAgent: result creation failed");
```

Comment:
need to decide what connector will be added to result if multiple connections between start/end exist